lte/gateway/c/oai/lib/pipelined_client/PipelinedServiceClient.h

```diff
@@ -0,0 +1,219 @@
+      uint32_t in_teid, const struct ip_flow_dl& flow_dl,
+      uint32_t flow_precedence, uint32_t state,
+      std::function<void(Status, UESessionContextResponse)> callback);
+
+  // APIs for paging IDLE -> PAGING, ACTIVATE -> Delete PAGING
+  static int UpdateUEIPv4SessionSet(
+      const struct in_addr& ue_ipv4_addr, uint32_t state,
+      std::function<void(grpc::Status, UESessionContextResponse)> callback);
+
+ public:
+  static PipelinedServiceClient& get_instance();
+  PipelinedServiceClient(PipelinedServiceClient const&) = delete;
+  void operator=(PipelinedServiceClient const&) = delete;
+
+  // Set the UE IPv4 address
+  static void ue_set_ipv4_addr(
+      const struct in_addr& ue_ipv4_addr, UESessionSet& request) {
+    IPAddress* encode_ue_ipv4_addr = request.mutable_ue_ipv4_address();
+    encode_ue_ipv4_addr->set_version(IPAddress::IPV4);
+    encode_ue_ipv4_addr->set_address(&ue_ipv4_addr, sizeof(struct in_addr));
+  }
+
+  // Set the UE IPv6 address
+  static void ue_set_ipv6_addr(
+      const struct in6_addr& ue_ipv6_addr, UESessionSet& request) {
+    IPAddress* encode_ue_ipv6_addr = request.mutable_ue_ipv6_address();
+    encode_ue_ipv6_addr->set_version(IPAddress::IPV6);
+    encode_ue_ipv6_addr->set_address(&ue_ipv6_addr, sizeof(struct in6_addr));
+  }
+
+  // Set the GNB IPv4 address
+  static void gnb_set_ipv4_addr(
+      const struct in_addr& gnb_ipv4_addr, UESessionSet& request) {
+    IPAddress* encode_gnb_ipv4_addr = request.mutable_enb_ip_address();
+    encode_gnb_ipv4_addr->set_version(IPAddress::IPV4);
+    encode_gnb_ipv4_addr->set_address(&gnb_ipv4_addr, sizeof(struct in_addr));
+  }
+
+  // Set the Session Config State
+  static void config_ue_session_state(
+      uint32_t& ue_state, UESessionSet& request) {
+    UESessionState* ue_session_state = request.mutable_ue_session_state();
+
+    if (UE_SESSION_ACTIVE_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::ACTIVE);
+    } else if (UE_SESSION_UNREGISTERED_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::UNREGISTERED);
+    } else if (UE_SESSION_INSTALL_IDLE_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::INSTALL_IDLE);
+    } else if (UE_SESSION_UNINSTALL_IDLE_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::UNINSTALL_IDLE);
+    } else if (UE_SESSION_SUSPENDED_DATA_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::SUSPENDED_DATA);
+    } else if (UE_SESSION_RESUME_DATA_STATE == ue_state) {
+      ue_session_state->set_ue_config_state(UESessionState::RESUME_DATA);
+    }
+  }
+
+  // Set the APN
+  static void config_apn(char* apn, UESessionSet& request) {
+    if (apn) {
+      request.set_apn(apn);
+    }
+  }
+
+ private:
+  PipelinedServiceClient();
+  std::unique_ptr<Pipelined::Stub> stub_{};
+  static const uint32_t RESPONSE_TIMEOUT = 30;  // seconds
```

Comment:
@koolzz can we confirm this is an acceptable timeout? 30 seconds for gRPC response seems high

Reply:
Changed it to 3. Is it fine.

Reply:
Yeah lets lower that, 5 seconds maybe?

Reply:
yeah looks like its 3 now that sounds fine